lasagne/layers/pool.py

```diff
@@ -15,81 +14,119 @@
+        must be True if pad is nonzero.
 
     Returns
     -------
     output_length
         * None if either input is None
-        * `input_length / pool_length` otherwise
+        * Computed length of the pooling operator otherwise
     '''
 
-    if input_length is None or pool_length is None:
+    if input_length is None or ds is None:
         return None
 
     if ignore_border:
-        # Round shape down when we ignore the border
-        return input_length // pool_length
+        output_length = input_length + 2 * pad - ds + 1
+        output_length = (output_length + stride - 1) // stride
 
-    # Otherwise, round the shape up
-    return int(np.ceil(float(input_length) / pool_length))
+    # output length calculation taken from:
+    # https://github.com/Theano/Theano/blob/master/theano/tensor/signal/downsample.py
+    else:
+        if stride >= ds:
+            output_length = (input_length + stride - 1) // stride
+        else:
+            output_length = max(
+                0, (input_length - ds + stride - 1) // stride) + 1
+
+    return output_length
 
 
 class MaxPool1DLayer(Layer):
-    def __init__(self, incoming, ds, ignore_border=False, **kwargs):
+
+    def __init__(self, incoming, ds, stride=None,
+                 ignore_border=False, pad=0, **kwargs):
         super(MaxPool1DLayer, self).__init__(incoming, **kwargs)
         self.ds = ds  # an integer
+        self.stride = ds if stride is None else stride
+        self.pad = pad
         self.ignore_border = ignore_border
 
     def get_output_shape_for(self, input_shape):
         output_shape = list(input_shape)  # copy / convert to mutable list
 
         output_shape[2] = pool_output_length(input_shape[2],
-                                             self.ds,
-                                             ignore_border=self.ignore_border)
+                                             ds=self.ds,
+                                             stride=self.stride,
+                                             ignore_border=self.ignore_border,
+                                             pad=self.pad,
+                                             )
 
         return tuple(output_shape)
 
     def get_output_for(self, input, **kwargs):
         input_4d = T.shape_padright(input, 1)
-        pooled = downsample.max_pool_2d(input_4d, (self.ds, 1),
-                                        self.ignore_border)
+        pooled = downsample.max_pool_2d(input_4d,
+                                        ds=(self.ds, 1),
+                                        st=(self.stride, 1),
+                                        ignore_border=self.ignore_border,
+                                        padding=(self.pad, 0),
+                                        )
         return pooled[:, :, :, 0]
 
 
 class MaxPool2DLayer(Layer):
-    def __init__(self, incoming, ds, ignore_border=False, **kwargs):
+
+    def __init__(self, incoming, ds, stride=None,
```

Comment:
In `MaxPool2DDNNLayer` the argument to specify the strides is called `strides` (because it is a tuple here). It's probably a good idea to make these interfaces consistent, either by changing it to `strides` here or changing it to `stride` in `MaxPool2DDNNLayer`. Personally I think `strides` makes more sense, as there are two of them in the 2D case.